Support configuration using a single Redis URL

martinmo opened the pull request:
This little PR adds support for specifying Redis connection details using a single URL, e.g. in combination with Django:

```python
HUEY = {
    'url': 'redis://localhost:6379/?db=1'
}
```

People wanting to implement [12factor configuration using envvars](https://12factor.net/config) will find this useful (at least I hope so).

The supported URL schemes are defined by redis-py and described in https://github.com/andymccurdy/redis-py/blob/master/redis/connection.py#L777.

coleifer commented on huey/storage.py:
```diff
@@ -122,6 +122,10 @@ def flush_all(self):
                  max_errors=1000, connection_pool=None, **connection_params):
+        redis_url = connection_params.pop('url', None)
+        if redis_url:
+            connection_pool = redis.ConnectionPool.from_url(redis_url, decode_components=True)
```
Is the `redis_url` mutually exclusive with other parameters here? What is the behavior if a user erroneously supplies arguments for both `url` and `connection_pool`?

coleifer replied:
Also, why is `url` not just a regular keyword argument?

`name='huey', blocking=False, read_timeout=1, max_errors=1000, connection_pool=None, url=None, **connection_params)`?

martinmo replied:
> Is the redis_url mutually exclusive with other parameters here? What is the behavior if a user erroneously supplies arguments for both url and connection_pool?

`redis_url` will take precedence even if a `connection_pool` or other `connection_params` have been specified. But such a conflict already existed with `connection_pool` and `connection_params`, where `connection_pool` wins if specified.

To fix this: Would you like a `ValueError` being raised if more than one of these parameters are specified at the same time? I'm happy to supplement this PR :)

> Also, why is url not just a regular keyword argument?

I just thought it was a good fit for `connection_params`.

coleifer replied:
> To fix this: Would you like a ValueError being raised if more than one of these parameters are specified at the same time? I'm happy to supplement this PR :)

That'd be awesome, do you mind? I can handle it as well ifyou prefer.

Making url a regular keyword argument also seems like a good idea to me.

martinmo commented:
I have incorporated your suggestions and also added some tests. Unfortunately on Python 2.6 they currently fail because I am obviously using `assertRaises()` in an backwards incompatible way.

martinmo commented:
Fixed :+1:

camilonova commented:
@martinmo please fix the conflicts to make easier merging it

martinmo commented:
@camilonova Done :+1:

@coleifer Waiting for your feedback :shipit:

camilonova commented:
@coleifer Please merge this one 🍔

coleifer commented on huey/storage.py:
```diff
@@ -104,8 +104,17 @@ def flush_all(self):
-                 max_errors=1000, connection_pool=None, **connection_params):
+    def __init__(self, name='huey', blocking=False, read_timeout=1, max_errors=1000,
+            connection_pool=None, url=None, **connection_params):
+        if sum(bool(x) for x in (url, connection_pool, connection_params)) > 1:
```
Can you change this to:

```python
if len(filter(None, (url, connection_pool, connection_params))) > 1:
```

martinmo replied:
No :( For compatibility with Python 3, `filter()` would need to be wrapped by `list()`:

```python
if len(list(filter(None, (url, connection_pool, connection_params)))) > 1:
```

IMHO this is a readability "downgrade" in comparison to the original code. If `len()` is a must have for you I would suggest using `len([x for x in (url, ...) if x]) > 1`, which works on all Pythons >= 2.6.

coleifer replied:
`len` is semantically what you are interested in. The use of `sum` is predicated on the strange fact that Python treats `True` as `1`.

coleifer commented on huey/storage.py:
```diff
@@ -104,8 +104,17 @@ def flush_all(self):
+        if sum(bool(x) for x in (url, connection_pool, connection_params)) > 1:
+            raise ValueError(
+                "The following arguments are mutually exclusive: "
+                "url, connection_pool and connection_params"
```
Can you change the error message to:

> The connection configuration is over-determined. Please specify only one of the
> the following: "url", "connection_pool", or "connection_params"

martinmo replied:
Sure.

coleifer commented on huey/storage.py:
```diff
@@ -104,8 +104,17 @@ def flush_all(self):
+        if url:
+            connection_pool = redis.ConnectionPool.from_url(url, decode_components=True)
+
         if connection_pool is None:
```
This can be an `elif`, right?

martinmo replied:
Yes.

coleifer commented:
A couple small little changes and then I'd be happy to merge this. It'd be a very nice feature to have, so thank you for taking the time to implement it and add tests. I appreciate it.

martinmo commented:
Thanks for the feedback! With one minor exception (see [this comment](#discussion_r74258087)), I have incorporated all suggested changes.

coleifer commented on huey/storage.py:
```diff
@@ -104,9 +104,18 @@ def flush_all(self):
-    def __init__(self, name='huey', blocking=False, read_timeout=1,
-                 max_errors=1000, connection_pool=None, **connection_params):
-        if connection_pool is None:
+    def __init__(self, name='huey', blocking=False, read_timeout=1, max_errors=1000,
```
For future reference please note that peewee uses a 79 character line limit. I will go ahead and fix it once it's merged.

coleifer commented on huey/storage.py:
```diff
@@ -104,9 +104,18 @@ def flush_all(self):
-        if connection_pool is None:
+    def __init__(self, name='huey', blocking=False, read_timeout=1, max_errors=1000,
+            connection_pool=None, url=None, **connection_params):
+        if sum(bool(x) for x in (url, connection_pool, connection_params)) > 1:
```
You're relying on the quirk that Python considers `True` to be `1` -- that is why this trick works. I think it is quite obscure, though...summing booleans.

Since `filter()` isn't as nice because fucking stupid ass Python3, how about changing this to a list comprehension?

```python
if len(item for item in (url, connection_pool, connection_params)) > 1:
```

Would you consider changing that?

coleifer commented on huey/tests/test_storage.py:
```diff
@@ -94,3 +97,27 @@ def test_event_iterator(self):
+
+    def test_conflicting_init_args3(self):
+        self.assertRaises(ValueError, RedisStorage,
+                          connection_pool=ConnectionPool(), url="redis://localhost")
```
You can reduce these into a single test using `itertools.combinations`:

```python
for pair in itertools.combinations((url, connection_pool, host), 2):
   self.assertRaises(ValueError, RedisStorage, ...)
```

That's not exactly what you need, but hopefully that makes sense?

coleifer commented:
Once you feel all comments are addressed, please also rebase this into a single commit. Thanks!

martinmo commented:
Done.

martinmo commented:
Cool. Thanks for merging! :tada:

> I'm not sure if the brackets are necessary

They are, you'll get a `TypeError: object of type 'generator' has no len()` on all Pythons.

coleifer commented:
Ahh, thanks for letting me know. I utterly despise Python 3!

Nice work on the patch, please feel free to let me know of any other ideas you have as I'd welcome more patches of that quality.

coleifer commented:
Oh, and btw, I finally settled on:

```python
if sum(1 for p in (url, connection_pool, connection_params) if p) > 1:
```

lol
